pymc/logprob/transforms.py

```diff
@@ -1229,6 +1232,10 @@ def transformed_logprob(op, values, *inputs, use_jacobian=True, **kwargs):
                 assert isinstance(value.owner.op, TransformedVariable)
                 original_forward_value = value.owner.inputs[1]
                 jacobian = transform.log_jac_det(original_forward_value, *inputs).copy()
+                # Check if jacobian dimensions align with logp
+                if jacobian.ndim < logp.ndim:
+                    diff_ndims = logp.ndim - jacobian.ndim
```

Comment:
I would love if you only had to test whether the `jacobian` and `logp` `ndims` were the same or slightly different. As far as I understand this is the key piece of the PR so I want to be extra sure that we're covering every angle. Could you tell me if it makes sense to take a defensive approach and test the following circumstances?

- `jacobian.ndim > logp.ndim`: an example would be a `Log` transform on an `MvNormal`. It doesn't really make sense to me, but it _could_ happen under some circumstances that I don't really know at the moment. What would be the approach to take there? To sum the across the jacobian extra dimensions?
- `jacobian.ndim <= logp.ndim` and `jacobian.shape != logp.shape[:jacobian.ndim]`: This would mean that the "batch" dimensions of the jacobian don't match with the "batch" dimensions of the `logp`. To me this sounds like an error and it should at some point. The question is if it makes sense to add an assertion to the `logp` graph or not. It might be a very picky thing to do, and completely wasteful.
- `jacobian.ndim > logp.ndim` and `jacobian.shape[:logp.ndim] != logp.shape`: to me, this is just like the case listed above. Something was wrong in the `jacobian` or the `logp` and an error should be raised. The question is whether it makes sense to do so at each `logp` evaluation.

If you can't argue why you don't need to test the above cases, then I think that this change is good enough

Reply:
Good points. 

> jacobian.ndim > logp.ndim: an example would be a Log transform on an MvNormal. It doesn't really make sense to me, but it could happen under some circumstances that I don't really know at the moment. What would be the approach to take there? To sum the across the jacobian extra dimensions?

That touches on https://github.com/pymc-devs/pymc/issues/6360#issuecomment-1621358657, and in generally the jacobian can't be simply summed if the core entries are somewhat "redundant". For instance, we shouldn't count the jacobian 3 times for a log transform dirichlet([1, 1, 1]), but we can do it for a multivariate normal. I will add an error for this.

> jacobian.ndim <= logp.ndim and jacobian.shape != logp.shape[:jacobian.ndim]: This would mean that the "batch" dimensions of the jacobian don't match with the "batch" dimensions of the logp. To me this sounds like an error and it should at some point. The question is if it makes sense to add an assertion to the logp graph or not. It might be a very picky thing to do, and completely wasteful.

The nice thing is that the static shape information is capturing more and more of these errors without any extra symbolic operations. Indeed this was what caused test failures when we bumped the PyTensor dependency in another PR and let me to realize our approach was completely wrong (even in the case it doesn't fail because it broadcasts). A correct jacobian determinant should never affect the batch dimensions, and shouldn't have "lingering" core dimensions. I think it's fine to assume they will be correctly implemented going into the future, and leave edge bugs to be captured by static shape info and making sure there is no sneaky broadcasting going on (which the changed test now does)